synapse/logging/_structured.py

```diff
@@ -0,0 +1,268 @@
+        """
+        Args:
+            observer (ILogObserver): A Twisted logging observer.
+            *args, **kwargs: Args/kwargs to be passed to logging.Handler.
+        """
+        self.observer = observer
+        super().__init__(*args, **kwargs)
+
+    def emit(self, record):
+        """
+        Emit a record to Twisted's observer.
+
+        Args:
+            record (logging.LogRecord)
+        """
+
+        self.observer(
+            {
+                "log_time": record.created,
+                "log_text": record.getMessage(),
+                "log_format": "{log_text}",
+                "log_namespace": record.name,
+                "log_level": stdlib_log_level_to_twisted(record.levelname),
+            }
+        )
+
+
+def SynapseFileLogObserver(outFile):
+    """
+    A log observer that formats events like the traditional log formatter and
+    sends them to `outFile`.
+
+    Args:
+        outFile (file object): The file object to write to.
+    """
+
+    def formatEvent(_event):
+        event = _event.copy()
+        event["log_level"] = event["log_level"].name.upper()
+        event["log_format"] = (
+            "- {log_namespace} - {log_level} - {request} - "
+            + event.get("log_format", "{log_text}")
+        )
+        return eventAsText(event, includeSystem=False) + "\n"
+
+    return FileLogObserver(outFile, formatEvent)
+
+
+class LoggingOutputType(Names):
+    CONSOLE = NamedConstant()
+    JSON = NamedConstant()
+    FLUENTD = NamedConstant()
+
+
+class OutputPipeType(Values):
+    stdout = ValueConstant(sys.__stdout__)
+    stderr = ValueConstant(sys.__stderr__)
+
+
+@attr.s
+class HandlerConfiguration(object):
+    name = attr.ib()
+    type = attr.ib()
+    location = attr.ib()
+
+
+def parse_handler_configs(config):
+    """
+    Parse the logging format version 2 handler section.
+
+    Args:
+        config (dict): A logging format v2 dictionary.
+
+    Yields:
+        HandlerConfiguration instances.
+
+    Raises:
+        ConfigError: If any of the handler configuration items are invalid.
+    """
+    for name, config in config.get("handlers").items():
+        logging_type = None
+
+        if "type" not in config and "class" in config:
+            # Handle the old "class": style.
+            if config["class"] == "logging.StreamHandler":
+                yield (
+                    HandlerConfiguration(
+                        name=name,
+                        type=LoggingOutputType.CONSOLE,
+                        location=sys.__stdout__,
+                    )
+                )
+            else:
+                raise ConfigError(
+                    "The logging class %s is not supported in logging format 2."
+                    % (config["class"],)
+                )
+        elif "type" in config:
+            logging_type = LoggingOutputType.lookupByName(config["type"].upper())
+
+            if logging_type in [LoggingOutputType.CONSOLE, LoggingOutputType.JSON]:
```

Comment:
shouldn't we reject other types, rather than ignore them?

Reply:
done